### DISCUSSION TITLE
fcitx5 does not work only on Firefox

### DISCUSSION THREAD

**Discussion by phanect** [fcitx/fcitx5#789] [Q&A] [answered]
CPU Architecture: x86_64
Distribution: KDEneon (an Ubuntu-based Linux distro)
Graphics Platform: X11
Desktop Environment: KDE
Firefox version: 113.0, installed from https://ppa.launchpadcontent.net/mozillateam/ppa/ubuntu (Mozilla PPA, not Ubuntu's default repository)
fcitx version: 5.0.23, installed from http://archive.neon.kde.org/user (KDE neon)

I want to input Japanese using fcitx5 and mozc.
On other apps, fcitx works fine and I can input Japanese without any problems. It works appropriately on Kate (Qt-based), Inkscape (GTK-based), and Google Chrome (GTK-based?).

However, I cannot input Japanese only on Firefox.

The required environment variables seem appropriately set:

```shell
$ echo $GTK_IM_MODULE
fcitx
$ echo $QT_IM_MODULE
fcitx
$ echo $XMODIFIERS
@im=fcitx
```

and fcitx5 seems to have started automatically:

```shell
$ ps aux | grep fcitx
phanect    24850  0.0  0.2 220552 36948 ?        Sl   08:18   0:00 /usr/bin/fcitx5 -r
phanect    30068  0.0  0.0  10256  2648 pts/1    S+   08:53   0:00 grep --color=auto fcitx
```

I found I can input Japanese properly if I set GTK_IM_MODULE=xim.

```shell
$ GTK_IM_MODULE=xim firefox
```

However, I hear XIM has some issues and not is recommended, so I want to know if I have different way to solve this problem.
In addition, I want to suggest changes to the packagers or developers of this Firefox or fcitx package so that you can use fcitx without setting environment variables manually.

My questions are:

1. Do I really have to set GTK_IM_MODULE=xim to use fcitx on Firefox, or is there any better solution?
2. Which package or software should be fixed to avoid appending an environment variable manually?
    - KDE neon's fcitx5 package(s)?
    - Firefox package in `mozillateam` PPA?
    - Firefox itself?
    - fcitx5 itself?

Thanks in advance,
Jumpei

**Comment by naokiri** [comment]
Well, the real problem is that GTK doesn't support any of the wayland zwp-input-method protocol so the answer for 2. should be the GTK...
But in my understanding latest fcitx5 have implemented the legacy(?) protocols that GTK expects for inputs so maybe the neon's fcitx5 haven't included those features on building?

Maybe pasting the output of `fcitx5-diagnose` help more to tell the situation.

**Reply by naokiri** [reply]
Sorry my comment was incorrect, it was not fcitx5 that implemented them but https://github.com/fcitx/fcitx5-gtk this module that does the GTK specific messaging. Please check the `fcitx5-diagnose`'s output anyway.

**Reply by phanect** [reply]
@naokiri Thanks for the reply!

> Well, the real problem is that GTK doesn't support any of the wayland zwp-input-method protocol so the answer for 2. should be the GTK...

I think I'm not using Wayland but X11 as I wrote. Is it still relevant to this problem?

> Maybe pasting the output of `fcitx5-diagnose` help more to tell the situation.

Thanks, here's the output of `fcitx5-diagnose`.

<details>
<summary>fcitx5-diagnose result</summary>

# System Info:
1.  `uname -a`:

        Linux phanect-neon-undefined 5.19.0-41-generic #42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

2.  `lsb_release -a`:

        No LSB modules are available.
        Distributor ID: Neon
        Description:    KDE neon 5.27
        Release:        22.04
        Codename:       jammy

3.  `lsb_release -d`:

        Description:    KDE neon 5.27

4.  `/etc/lsb-release`:

        DISTRIB_ID=neon
        DISTRIB_RELEASE=22.04
        DISTRIB_CODENAME=jammy
        DISTRIB_DESCRIPTION="KDE neon 5.27"

5.  `/etc/os-release`:

        PRETTY_NAME="KDE neon 5.27"
        NAME="KDE neon"
        VERSION_ID="22.04"
        VERSION="5.27"
        VERSION_CODENAME=jammy
        ID=neon
        ID_LIKE="ubuntu debian"
        HOME_URL="https://neon.kde.org/"
        SUPPORT_URL="https://neon.kde.org/"
        BUG_REPORT_URL="https://bugs.kde.org/"
        PRIVACY_POLICY_URL="https://kde.org/privacypolicy/"
        UBUNTU_CODENAME=jammy

6.  Desktop Environment:

    Desktop environment is `kde`.

7.  XDG SESSION TYPE:

        XDG_SESSION_TYPE='x11'

8.  Bash Version:

        BASH_VERSION='5.1.16(1)-release'

# Environment:
1.  DISPLAY:

        DISPLAY=':0'


        WAYLAND_DISPLAY=''

2.  Keyboard Layout:

    1.  `setxkbmap`:

            xkb_keymap {
                xkb_keycodes  { include "evdev+aliases(qwerty)" };
                xkb_types     { include "complete"      };
                xkb_compat    { include "complete+japan"        };
                xkb_symbols   { include "pc+jp+inet(evdev)"     };
                xkb_geometry  { include "pc(pc105)"     };
            };

    2.  `xprop`:

            _XKB_RULES_NAMES(STRING) = "evdev", "pc105", "jp", "", ""

3.  Locale:

    1.  All locales:

            C
            C.utf8
            en_AG
            en_AG.utf8
            en_AU.utf8
            en_BW.utf8
            en_CA.utf8
            en_DK.utf8
            en_GB.utf8
            en_HK.utf8
            en_IE.utf8
            en_IL
            en_IL.utf8
            en_IN
            en_IN.utf8
            en_NG
            en_NG.utf8
            en_NZ.utf8
            en_PH.utf8
            en_SG.utf8
            en_US.utf8
            en_ZA.utf8
            en_ZM
            en_ZM.utf8
            en_ZW.utf8
            ja_JP.utf8
            POSIX

    2.  Current locale:

            LANG=en_US.UTF-8
            LANGUAGE=en_US
            LC_CTYPE="en_US.UTF-8"
            LC_NUMERIC=ja_JP.UTF-8
            LC_TIME=ja_JP.UTF-8
            LC_COLLATE="en_US.UTF-8"
            LC_MONETARY=ja_JP.UTF-8
            LC_MESSAGES="en_US.UTF-8"
            LC_PAPER=ja_JP.UTF-8
            LC_NAME=ja_JP.UTF-8
            LC_ADDRESS=ja_JP.UTF-8
            LC_TELEPHONE=ja_JP.UTF-8
            LC_MEASUREMENT=ja_JP.UTF-8
            LC_IDENTIFICATION=ja_JP.UTF-8
            LC_ALL=

4.  Directories:

    1.  Home:

            /home/phanect

    2.  `${XDG_CONFIG_HOME}`:

        Environment variable `XDG_CONFIG_HOME` is not set.

        Current value of `XDG_CONFIG_HOME` is `~/.config` (`/home/phanect/.config`).

    3.  Fcitx5 Settings Directory:

        Current fcitx5 settings directory is `~/.config/fcitx5` (`/home/phanect/.config/fcitx5`).

5.  Current user:

    The script is run as phanect (1000).

# Fcitx State:
1.  executable:

    Found fcitx5 at `/usr/bin/fcitx5`.

2.  version:

    Fcitx version: `5.0.23`

3.  process:

    Found 1 fcitx5 process:

           1598 fcitx5

4.  `fcitx5-remote`:

    `fcitx5-remote` works properly.

5.  DBus interface:

    Using `dbus-send` to check dbus.

    Owner of DBus name `org.fcitx.Fcitx5` is `:1.5`.

    PID of DBus name `org.fcitx.Fcitx5` owner is `1598`.

    Debug information from dbus:

           Group [x11::0] has 3 InputContext(s)
          IC [e3f880ae45194bdf9e06153f6c836505] program:yakuake frontend:dbus cap:6000820872 focus:1
          IC [dc12827e7230498f9ec38ccbfcb1f7cd] program:plasmashell frontend:dbus cap:16000800072 focus:0
          IC [5ac23dfe4b6241d9ae48896ed5addc1f] program:plasmashell frontend:dbus cap:16000800072 focus:0
        Input Context without group

# Fcitx Configure UI:
1.  Config Tool Wrapper:

    Found fcitx5-configtool at `/usr/bin/fcitx5-configtool`.

2.  Config GUI for qt:

    Found `fcitx5-config-qt` at `/usr/bin/fcitx5-config-qt`.

3.  Config GUI for kde:

    Found fcitx5 kcm module.

        kcm_fcitx5                     - Configure Input Method

# Frontends setup:
## Xim:
1.  `${XMODIFIERS}`:

    Environment variable XMODIFIERS is set to "@im=fcitx" correctly.
    Xim Server Name from Environment variable is fcitx.

2.  XIM_SERVERS on root window:

    Xim server name is the same with that set in the environment variable.

## Qt:
1.  qt4 - `${QT4_IM_MODULE}`:

    Environment variable QT_IM_MODULE is set to "fcitx" correctly.

2.  qt5 - `${QT_IM_MODULE}`:

    Environment variable QT_IM_MODULE is set to "fcitx" correctly.

3.  Qt IM module files:

    Found fcitx5 qt5 module: `/lib/x86_64-linux-gnu/fcitx5/qt5/libfcitx-quickphrase-editor5.so`.
    Found fcitx5 im module for qt6: `/lib/x86_64-linux-gnu/qt6/plugins/platforminputcontexts/libfcitx5platforminputcontextplugin.so`.
    Found fcitx5 im module for qt5: `/lib/x86_64-linux-gnu/qt5/plugins/platforminputcontexts/libfcitx5platforminputcontextplugin.so`.
    Found unknown fcitx qt module: `/lib/x86_64-linux-gnu/qt5/plugins/kcms/kcm_fcitx5.so`.

    Following error may not be accurate because guessing Qt version from path depends on how your distribution packages Qt. It is not a critical error if you do not use any Qt application with certain version of Qt or you are using text-input support by Qt under Wayland.
    **Cannot find fcitx5 input method module for Qt4.**

## Gtk:
1.  gtk - `${GTK_IM_MODULE}`:

    Environment variable GTK_IM_MODULE is set to "fcitx" correctly.

2.  `gtk-query-immodules`:

    1.  gtk 2:

        **Cannot find `gtk-query-immodules` for gtk 2**

        **Cannot find fcitx5 im module for gtk 2.**

    2.  gtk 3:

        **Cannot find `gtk-query-immodules` for gtk 3**

        **Cannot find fcitx5 im module for gtk 3.**

3.  Gtk IM module cache:

    1.  gtk 2:

        Found immodules cache for gtk `2.24.33` at `/lib/x86_64-linux-gnu/gtk-2.0/2.10.0/immodules.cache`.
        Version Line:

            # Created by /usr/lib/x86_64-linux-gnu/libgtk2.0-0/gtk-query-immodules-2.0 from gtk+-2.24.33

        Found fcitx5 im modules for gtk `2.24.33`.

            "/usr/lib/x86_64-linux-gnu/gtk-2.0/2.10.0/immodules/im-fcitx5.so" 
            "fcitx" "Fcitx5 (Flexible Input Method Framework5)" "fcitx5" "/usr/locale" "ja:ko:zh:*" 
            "fcitx5" "Fcitx5 (Flexible Input Method Framework5)" "fcitx5" "/usr/locale" "ja:ko:zh:*" 

    2.  gtk 3:

        Found immodules cache for gtk `3.24.33` at `/lib/x86_64-linux-gnu/gtk-3.0/3.0.0/immodules.cache`.
        Version Line:

            # Created by /usr/lib/x86_64-linux-gnu/libgtk-3-0/gtk-query-immodules-3.0 from gtk+-3.24.33

        Found fcitx5 im modules for gtk `3.24.33`.

            "/usr/lib/x86_64-linux-gnu/gtk-3.0/3.0.0/immodules/im-fcitx5.so" 
            "fcitx" "Fcitx5 (Flexible Input Method Framework5)" "fcitx5" "/usr/locale" "ja:ko:zh:*" 
            "fcitx5" "Fcitx5 (Flexible Input Method Framework5)" "fcitx5" "/usr/locale" "ja:ko:zh:*" 

    3.  gtk 4:

        **Cannot find immodules cache for gtk 4**

        **Cannot find fcitx5 im module for gtk 4 in cache.**

4.  Gtk IM module files:

    1.  gtk 2:

        All found Gtk 2 immodule files exist.

    2.  gtk 3:

        All found Gtk 3 immodule files exist.

    3.  gtk 4:

        All found Gtk 4 immodule files exist.

# Configuration:
## Fcitx Addons:
1.  Addon Config Dir:

    Found fcitx5 addon config directory: `/usr/share/fcitx5/addon`.

2.  Addon List:

    1.  Found 20 enabled addons:

            Classic User Interface 5.0.23
            Clipboard 5.0.23
            DBus 5.0.23
            DBus Frontend 5.0.23
            Emoji 5.0.23
            Fcitx4 Frontend 5.0.23
            IBus Frontend 5.0.23
            Input method selector 5.0.23
            Keyboard 5.0.23
            KDE Input Method Panel 5.0.23
            Mozc 
            Status Notifier 5.0.23
            Notification 5.0.23
            Quick Phrase 5.0.23
            Spell 5.0.23
            Unicode 5.0.23
            Wayland 5.0.23
            Wayland Input method frontend 5.0.23
            XCB 5.0.23
            X Input Method Frontend 5.0.23

    2.  Found 0 disabled addons:

3.  Addon Libraries:

    All libraries for all addons are found.

4.  User Interface:

    Found 2 enabled user interface addons:

        Classic User Interface
        KDE Input Method Panel

    Kimpanel process:

           1740 /usr/bin/plasmashell --no-respawn

## Input Methods:
1.  `/home/phanect/.config/fcitx5/profile`:

        [Groups/0]
        # Group Name
        Name=デフォルト        # Layout
        Default Layout=jp
        # Default Input Method
        DefaultIM=mozc

        [Groups/0/Items/0]
        # Name
        Name=mozc
        # Layout
        Layout=jp

        [GroupOrder]
        0=デフォルト
# Log:
1.  `date`:

        2023年  5月 11日 木曜日 11:36:05 JST

2.  `/home/phanect/.config/fcitx5/crash.log`:

    `/home/phanect/.config/fcitx5/crash.log` not found.

**Warning: the output of fcitx5-diagnose contains sensitive information, including the distribution name, kernel version, name of currently running programs, etc.**

**Though such information can be helpful to developers for diagnostic purpose, please double check and remove as necessary before posting it online publicly.**
</details>

I noticed gtk-query-immodules was not found according to `fcitx5-diagnose`, but I'm not sure how can I resolve this problem.
gtk-query-immodules-3.0 seems included in libgtk-3-bin package, but libgtk-3-bin package is already installed on my end.

**Reply by naokiri** [reply]
From the diagnose result, Gtk IM module cache exists with fcitx entry so the gtk-query-immodules won't be the issue. GTK apps will read the cache only.

I tried KDEneon on a VM, ran `fcitx5 -d --verbose=*=5` and checked the message. On firefox, deactivateInputMethod was called. There were no errors on fcitx, so it seems like firefox is deactivating IM on purpose.

I tested with geany (another GTK based app). There were no problem using fcitx with Japanese in geany.

So, this problem seems to be the problem of firefox of that build. Sorry I don't know the detail in the firefox so I cannot help further.

I personally use firefox nightly self build on Ubuntu 22.04 with fcitx5 and having no problem to input Japanse, so self-building might help your env too.

**Reply by phanect** [reply]
@naokiri Thanks for the investigation!
I understand it may be a problem with Firefox.
Since fcitx properly works on your self-build, it looks like a problem with Firefox's deb package.
I will contact the developer who made the package first.

**Reply by phanect** [reply]
I tried to install Ubuntu's default Firefox package (snap) and I could properly type Japanese with fcitx5 on it.
This looks like a problem in the Firefox deb package from the ppa.

**Reply by phanect** [reply]
@naokiri I have asked in Ubuntu Japanese mailing list and one of the members Shibata-san has investigated for me.

Cannot input Japanese using fcitx5 on Firefox installed from the deb from the PPA (Japanese)
https://lists.ubuntu.com/archives/ubuntu-jp/2023-May/006616.html

It looks like caused by an AppArmor profile for Firefox. It may prevent the connection between Firefox and fcitx via dbus.
By disabling the Firefox profile of AppArmor with the following command, I could successfully input Japanese using fcitx5.

$ sudo ln -s /etc/apparmor.d/usr.bin.firefox /etc/apparmor.d/disable/
$ sudo apparmor_parser -R /etc/apparmor.d/usr.bin.firefox

According to Shibata-san, Kubuntu 22.04 LTS has disabled this AppArmor profile so that fcitx properly works on Firefox.
I know it is not the best idea to disable the profile because it may degrade the security, but I don't have better ideas for now, so I will ask KDE neon developers to disable the profile.

Thank you for your help to investigate this issue!

**Comment by MrElectronics** [comment]
Sorry for being late to the party. In case anybody drops on this while investigating for a fix. Please consider to not trade in security (disable apparmor profile completely) for basic features. Simply add:
```
--- /etc/apparmor.d/usr.bin.firefox.orig	2024-08-09 05:53:58.268654820 +0200
+++ /etc/apparmor.d/usr.bin.firefox	2024-08-09 05:45:15.448550947 +0200
@@ -20,6 +20,7 @@
   include <abstractions/dbus-session-strict>
   include <abstractions/dbus-strict>
   include <abstractions/dconf>
+  include <abstractions/fcitx>
   include <abstractions/gnome>
   include <abstractions/ibus>
   include <abstractions/nameservice>
```
into apparmors firefox profile `/etc/apparmor.d/usr.bin.firefox` - obviously in mine ibus was already present while fcitx was missing. Or simply start firefox ones, try to use fcitx and let `aa-logprof` do the heavy weight lifting of digging out suggesting the parts obviously missing to the apparmor profile for proper functionality. After thoroughly reviewing your individual config for any it security vulnerabilities this might imply you should be good to go. Sincerely hope this finds you well, よろしくお願いいたします

**Reply by phanect** [reply, marked as answer]
Thanks for the info, @MrElectronics! ありがとうございます!

In fact, this problem seems fixed on KDE neon after the KDE 6 has been released on the last February.
I clean-installed KDE neon after the release, and I haven't experienced this issue on my end since then.

Note for the users of other Linux distributions: See [the parent post of this post](https://github.com/fcitx/fcitx5/discussions/789#discussioncomment-10281928).
